Modify Copy to automatically lock sync.Lockers

jbardin opened the pull request:
Whenever a value which is also a sync.Locker is encountered during Copy,
and it isn't specifically a sync.Mutex or sync.RWMutex, we lock it while
copying any part of that value.

mitchellh commented on copystructure.go:
```diff
@@ -48,14 +49,25 @@ type walker struct {
+	// space to use depth as an index
+	for w.depth >= len(w.locks) {
+		w.locks = append(w.locks, nil)
+	}
```
You can use `cap` here to do this w/o the `append` magic since the zero value of this slice is going to be `nil` anyways.

jbardin replied:
Sorry, I'm not sure I follow. This is the only place `w.locks` is extended to match `w.depth` as we walk deeper, so it would still need to be append at some point. All other access is indexed directly.

mitchellh replied:
When I actually coded it out your solution is actually a lot shorter, so its probably fine at the expense of more CPU cycles probably but we don't need to optimize for that here.

jbardin replied:
Oh, not sure if my comment may have thrown you off. I wrote that in an iteration where `w.locks` was being resliced to pop off values, but now it's never truncated. This only needs to ensure it's long enough to index with `w.depth` at any point.

mitchellh commented:
This looks pretty great. My feedback on how we should do this is to expand the API as follows:

  * Introduce a struct named something like `Copier` that has a `Copy` method on it
  * Top-level `Copy` becomes shorthand for `&Copier{}.Copy()`
  * Field on `Copier` such as `Lock bool` that if true will lock all lockers.

This lets us keep BC and introduce a new struct we can use for options and tweaking of the behavior of Copystructure.

jbardin commented:
Added an `Opts` type with a `Copy` method. Maybe there's a better name than Opts, but `Copier` is too close to `Copiers` and isn't the type contained in that map. 

Also added the ability to take Rlocks from an `RLocker()` method.

mitchellh commented:
Makes sense @jbardin. To bike shed: let's name it `Config`, that lines up more with Go standard libraries. Everything else looks good!

jbardin commented:
s/Opts/Config/

mitchellh commented:
This looks good to me! It has a conflict now :) Once its done I can merge it.

jbardin commented:
That was a rough rebase for some reason 😓 . Too many matching lines in the merges completely broke the patches apart.

Anyways, look like it should be OK to merge now.

mitchellh commented on copystructure.go:
```diff
@@ -2,27 +2,19 @@ package copystructure
 
-	return result, nil
+// LockedCopy returns a deep copy of v, taking locks as needed during the walk.
+func LockedCopy(v interface{}) (interface{}, error) {
```
One more nitpick: I think we just remove this and if you want locks then you use the `Config` structure.

mitchellh commented:
Sorry one more thing and then we're okay 😓

jbardin commented:
Done!

mitchellh commented:
Thanks!
